activemodel/lib/active_model/type/helpers/time_value.rb

```diff
@@ -22,7 +22,7 @@ def serialize(value)
         end
 
         def apply_seconds_precision(value)
-          return value unless precision && value.respond_to?(:usec)
+          return value unless precision && value.respond_to?(:usec) && value.usec.positive?
```

Comment:
The most slowest method call is the `value.change`.
We could also safe to bypass the line if `precision == 6`.

```ruby
if precision && precision < 6 && value.respond_to?(:usec) && value.usec.positive?
  ...
else
  value
end
```

Reply:
Indeed. Updated w/ the additional check for precision >= 6. Here is the updated benchmark used to run the results in my second commit:

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "benchmark-ips"
end

require "active_model"

# Current code for ActiveModel::Type::Helpers::TimeValue#apply_seconds_precision
def apply_precision_old(precision, value)
  return value unless precision && value.respond_to?(:usec)
  number_of_insignificant_digits = 6 - precision
  round_power = 10**number_of_insignificant_digits
  value.change(usec: value.usec - value.usec % round_power)
end

# Propposed code for ActiveModel::Type::Helpers::TimeValue#apply_seconds_precision
# that checks for positive usec before doing further calculations
def apply_precision_new(precision, value)
  return value unless precision && precision < 6 && value.respond_to?(:usec) && value.usec.positive?
  number_of_insignificant_digits = 6 - precision
  round_power = 10**number_of_insignificant_digits
  value.change(usec: value.usec - value.usec % round_power)
end

# Enumerate some representative scenarios here.
#
# It is very easy to make an optimization that improves performance for a
# specific scenario you care about but regresses on other common cases.
# Therefore, you should test your change against a list of representative
# scenarios. Ideally, they should be based on real-world scenarios extracted
# from production applications.
SCENARIOS = {
  "Time - No Precision, No Milliseconds"            => [0, Time.utc(2019,1,1,12,0,1)],
  "Time - No Precision, Includes Milliseconds"      => [0, Time.utc(2019,1,1,12,0,1,5123)],
  "Time - Include Mid Precision, No Milliseconds"       => [3, Time.utc(2019,1,1,12,0,1)],
  "Time - Include Mid Precision, Includes Milliseconds" => [3, Time.utc(2019,1,1,12,0,1,5123)],
  "Time - Include Max Precision, No Milliseconds"       => [6, Time.utc(2019,1,1,12,0,1)],
  "Time - Include Max Precision, Includes Milliseconds" => [6, Time.utc(2019,1,1,12,0,1,5123)],
}

SCENARIOS.each_pair do |name, args|
  puts
  puts " #{name} ".center(80, "=")
  puts

  precision, time = args

  Benchmark.ips do |x|
    x.report("old?") { apply_precision_old(precision, time.dup) }
    x.report("new?") { apply_precision_new(precision, time.dup) }
    x.compare!
  end
end
```